contracts/interchain-token-service/src/abi.rs

```diff
@@ -0,0 +1,503 @@
+                message: message.abi_encode(env).to_alloc_vec().into(),
+            }
+            .abi_encode_params(),
+            Self::ReceiveFromHub(types::ReceiveFromHub {
+                source_chain,
+                message,
+            }) => ReceiveFromHub {
+                messageType: U256::from(types::MessageType::ReceiveFromHub as u32),
+                source_chain: soroban_string_to_string(source_chain),
+                message: message.abi_encode(env).to_alloc_vec().into(),
+            }
+            .abi_encode_params(),
+        };
+        Bytes::from_slice(env, &msg)
+    }
+
+    pub fn abi_decode(env: &Env, payload: &Bytes) -> Result<Self, MessageError> {
+        ensure!(payload.len() >= 32, MessageError::InsufficientMessageLength);
+
+        let binding = payload.to_alloc_vec();
+        let payload_array = binding.as_slice();
+
+        let message_type = MessageType::abi_decode(&payload_array[0..32], true)
+            .map_err(|_| MessageError::InvalidMessageType)?;
+
+        let message = match message_type {
+            MessageType::SendToHub => {
+                let decoded = SendToHub::abi_decode_params(payload_array, true)
+                    .map_err(|_| MessageError::AbiDecodeFailed)?;
+
+                Self::SendToHub(types::SendToHub {
+                    destination_chain: String::from_str(env, &decoded.destination_chain),
+                    message: Message::abi_decode(
+                        env,
+                        &Bytes::from_slice(env, decoded.message.as_ref()),
+                    )?,
+                })
+            }
+            MessageType::ReceiveFromHub => {
+                let decoded = ReceiveFromHub::abi_decode_params(payload_array, true)
+                    .map_err(|_| MessageError::AbiDecodeFailed)?;
+
+                Self::ReceiveFromHub(types::ReceiveFromHub {
+                    source_chain: String::from_str(env, &decoded.source_chain),
+                    message: Message::abi_decode(
+                        env,
+                        &Bytes::from_slice(env, decoded.message.as_ref()),
+                    )?,
+                })
+            }
+            _ => return Err(MessageError::InvalidMessageType),
+        };
+
+        Ok(message)
+    }
+}
+
+fn soroban_string_to_string(soroban_string: String) -> StdString {
+    let length = soroban_string.len() as usize;
+    let mut bytes = vec![0u8; length];
+    soroban_string.copy_into_slice(&mut bytes);
+    StdString::from_utf8(bytes).expect("Invalid UTF-8 sequence")
+}
+
+fn convert_to_i128(decoded_value: Uint<256, 4>) -> Result<i128, MessageError> {
+    let slice = decoded_value.as_le_slice();
+
+    if !slice[16..].iter().all(|&b| b == 0) {
+        return Err(MessageError::InvalidAmount);
+    }
+
+    let mut truncated = [0u8; 16];
+    truncated.copy_from_slice(&slice[..16]);
+
+    Ok(i128::from_le_bytes(truncated))
```

Comment:
```suggestion
    Ok(i128::from_le_bytes(value_bytes[..16]))
```

Reply:
@milapsheth I've been unable to make this conversion work without using slices. Do we need this to work with an owned value `[u8; 32]`?

Reply:
nah, it's fine if we can't get it to work